Signal/src/Models/SignalAttachment.swift

```diff
@@ -60,21 +60,26 @@ class SignalAttachment: NSObject {
 
     // MARK: Properties
 
-    let data: Data
+    let dataSource: DataSource
 
-    internal var temporaryDataUrl: URL?
+    public var data: Data {
+        return dataSource.data()
+    }
+    public var dataLength: UInt {
+        return dataSource.dataLength()
+    }
+    public var dataUrl: URL? {
+        return dataSource.dataUrl()
+    }
+    public var sourceFilename: String? {
+        return dataSource.sourceFilename()
+    }
 
     // Attachment types are identified using UTIs.
     //
     // See: https://developer.apple.com/library/content/documentation/Miscellaneous/Reference/UTIRef/Articles/System-DeclaredUniformTypeIdentifiers.html
     let dataUTI: String
 
-    // An optional field that indicates the filename, if known, for this attachment.
-    let filename: String?
```

Comment:
These moved into `DataSource`. We used this property to preserve filenames, e.g. when sending documents.